### PR TITLE
Update global.json to use .NET 8

### PR DESCRIPTION
**Pull request by m-redding**
Updating global.json to require .NET 8 for dev, and updating pipelines to run using .net 8

Not including any target framework or testing target updates. That is a pretty large/disruptive change and there isn't a need for it right now.

### CODE REVIEW

File: sdk/maintenance/Azure.ResourceManager.Maintenance/tests/Scenario/MaintanenceConfigurationCancellationTests.cs [hunk 1]
```diff
@@ -81,7 +81,7 @@ public async Task MaintenanceConfigurationCancelTest()
                     }
                     else
                     {
-                        throw ex;
```
Comment (m-redding) [inline sdk/maintenance/Azure.ResourceManager.Maintenance/tests/Scenario/MaintanenceConfigurationCancellationTests.cs:84]: Not sure why this wasn't caught before, this warning was added in .NET 5 - https://learn.microsoft.com/en-us/dotnet/core/compatibility/code-analysis/5.0/ca2200-rethrow-to-preserve-stack-details

File: sdk/core/Azure.Core/tests/compatibility/ExpectedAotWarnings.txt [hunk 2]
```diff
@@ -12,15 +14,26 @@ Azure\.Core.src.Shared.DiagnosticScope\.cs\(\d*\): AOT analysis warning IL3050:
...
-System\.Linq\.Expressions.src.System.Linq.Expressions.BinaryExpression\.cs\(\d*\): AOT analysis warning IL3050: System.Linq.Expressions.Expression.GetResultTypeOfShift\(Type,Type\): Using member 'System.Type.MakeGenericType\(Type\[\]\)' which has 'RequiresDynamicCodeAttribute'
-System\.Net\.Http.src.System.Net.Http.DiagnosticsHandler\.cs\(\d*\): AOT analysis warning IL3050: System\.Net\.Http\.DiagnosticsHandler\.Write<T>\(DiagnosticSource,String,!!0\): Using member 'System\.Diagnostics\.DiagnosticSource\.Write\(String,Object\)' which has 'RequiresDynamicCodeAttribute'
\ No newline at end of file
+Azure\.Core.src.DynamicData.DynamicData\.DynamicMetaObject\.cs\(\d*\): Trim analysis warning IL2026: Azure\.Core\.Serialization\.DynamicData\.MetaObject\.BindGetIndex\(GetIndexBinder,DynamicMetaObject\[\]\): Using member 'Azure\.Core\.Serialization\.DynamicData\.GetViaIndexerMethod' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code
```
Comment (m-redding) [inline sdk/core/Azure.Core/tests/compatibility/ExpectedAotWarnings.txt:24]: Adding these just to get the pipeline green. Will follow up - https://github.com/Azure/azure-sdk-for-net/issues/44296

File: sdk/core/Azure.Core/tests/compatibility/ExpectedAotWarnings.txt [hunk 3]
```diff
@@ -1,26 +1,36 @@
...
-System\.Linq\.Expressions.src.System.Linq.Expressions.BinaryExpression\.cs\(\d*\): AOT analysis warning IL3050: System.Linq.Expressions.Expression.GetResultTypeOfShift\(Type,Type\): Using member 'System.Type.MakeGenericType\(Type\[\]\)' which has 'RequiresDynamicCodeAttribute'
-System\.Net\.Http.src.System.Net.Http.DiagnosticsHandler\.cs\(\d*\): AOT analysis warning IL3050: System\.Net\.Http\.DiagnosticsHandler\.Write<T>\(DiagnosticSource,String,!!0\): Using member 'System\.Diagnostics\.DiagnosticSource\.Write\(String,Object\)' which has 'RequiresDynamicCodeAttribute'
\ No newline at end of file
+ILC : Trim analysis warning IL2026: Azure\.Core\.Serialization\.DynamicData: Using member 'Azure\.Core\.Serialization\.DynamicData\.DynamicDataJsonConverter\.DynamicDataJsonConverter\(\)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code\. Using DynamicData or DynamicDataConverter is not compatible with trimming due to reflection-based serialization
```
Comment (m-redding) [inline sdk/core/Azure.Core/tests/compatibility/ExpectedAotWarnings.txt:1]: Opened an issue to come back and clean these up: https://github.com/Azure/azure-sdk-for-net/issues/44296

File: eng/Directory.Build.Common.props [hunk 4]
```diff
@@ -78,6 +78,11 @@
     </NoWarn>
 
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <!-- 
+      .NET 8 now produces security warnings for nuget packages in dotnet restore. Keep these as warnings to avoid needing to update T1 library dependencies.
```
Comment (annelo-msft) [inline eng/Directory.Build.Common.props:82]: Is there a way to scope them to warnings for T1 libraries but errors for T2 libraries?  Mostly just curious.

Comment (m-redding) [reply eng/Directory.Build.Common.props:82]: Hmm yeah it seems like there would be a way to do that, let me see if I can figure it out

**Review by jsquire — APPROVED** [review]

File: eng/pipelines/templates/steps/install-dotnet.yml [hunk 5]
```diff
@@ -18,6 +18,14 @@ steps:
...
       performMultiLevelLookup: true
   - task: UseDotNet@2
+    displayName: 'Use .NET 7.0 SDK'
+    retryCountOnTaskFailure: 3
+    inputs:
+      # AspNetCore runtime pack can't be installed outside of SDK and we need it for integration tests
+      packageType: sdk
+      performMultiLevelLookup: true
+      version: "7.0.x"
+  - task: UseDotNet@2 # We need both 6 and 8 because the azure-sdk-tools repo only uses 6, so all tools from there fail without it
```
[the remark this answers is not in the document]
Comment (m-redding) [reply eng/pipelines/templates/steps/install-dotnet.yml:28]: Sorry will update to 6, 7, and 8. 8 is downloaded in L14 from the global json. We need 7 to run the .NET 7 tests in the pipelines, and 6 for both the .NET 6 tests and the tools.

File: eng/Directory.Build.Common.props [hunk 6]
```diff
@@ -78,6 +78,11 @@
...
+      See: https://learn.microsoft.com/en-us/dotnet/core/compatibility/sdk/8.0/dotnet-restore-audit#recommended-action
+    -->
+    <WarningsNotAsErrors>NU1901;NU1902;NU1903;NU1904</WarningsNotAsErrors>
```
[the remark this answers is not in the document]
Comment (m-redding) [reply eng/Directory.Build.Common.props]: I conditioned it on IsClientLibrary, which I got from here: https://github.com/Azure/azure-sdk-for-net/blob/main/eng/Packages.Data.props#L13 but I'm not 100% sure if that's the right condition. Should it be IsShippingClientLibrary?

[the remark this answers is not in the document]
Comment (jsquire) [reply eng/Directory.Build.Common.props]: `IsClientLibrary` will also potentially flag test support projects, perf projects, and samples projects.   The latter of those seems important.   Let's go with `IsClientLibrary` for now and we can adjust if we see issues.

**Pull request merged**